e2e/node/request-metadata.test.ts

```diff
@@ -0,0 +1,265 @@
+// subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
+// INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
+// PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
+// HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+// SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+//
+
+import {
+  afterAll,
+  beforeAll,
+  describe,
+  expect,
+  it,
+  test,
+  jest,
+} from "@jest/globals";
+
+import type { Session } from "@inrupt/solid-client-authn-node";
+import {
+  getNodeTestingEnvironment,
+  getPodRoot,
+  getAuthenticatedSession,
+} from "@inrupt/internal-test-env";
+import { getSolidDataset } from "../../src";
+
+const { Headers } = global;
+
+const env = getNodeTestingEnvironment();
+if (
+  env.features?.APPLICATION_DEFINED_REQUEST_METADATA === "false" ||
+  env.features?.APPLICATION_DEFINED_REQUEST_METADATA === ""
+) {
+  // eslint-disable-next-line jest/no-focused-tests, jest/expect-expect
+  test.only(`Skipping unsupported application-defined request metadata tests in ${env.environment}`, () => {});
+}
+
+if (
+  env.features?.APPLICATION_DEFINED_REQUEST_METADATA === "true" &&
+  typeof env.features?.APPLICATION_DEFINED_REQUEST_METADATA_HEADERS !== "string"
+) {
+  throw new Error(
+    "Missing E2E_TEST_FEATURE_APPLICATION_DEFINED_REQUEST_METADATA_HEADERS environment variable.",
+  );
+}
+
+const headers: Record<string, string>[] = [];
+if (
+  typeof env.features?.APPLICATION_DEFINED_REQUEST_METADATA_HEADERS === "string"
+) {
+  try {
+    const rawHeaders = JSON.parse(
+      env.features?.APPLICATION_DEFINED_REQUEST_METADATA_HEADERS,
+    );
+    for (const [key, value] of Object.entries(rawHeaders)) {
+      if (typeof value !== "string") {
+        throw new Error();
+      }
+      headers.push({ [key]: value });
+    }
+  } catch {
+    throw new Error(
+      `E2E_TEST_FEATURE_APPLICATION_DEFINED_REQUEST_METADATA_HEADERS is malformed: expected a JSON key:value map, found ${env.features?.APPLICATION_DEFINED_REQUEST_METADATA_HEADERS}`,
+    );
+  }
+}
+
+const customizeHeaders = (
+  customizedFetch: typeof fetch,
+  requestHeaders: Record<string, string>,
+  responseHeaders: (h: Headers) => void,
+) => {
+  return async (
+    info: Parameters<typeof fetch>[0],
+    init?: Parameters<typeof fetch>[1],
+  ) => {
+    const response = await customizedFetch(info, {
+      ...init,
+      headers: {
+        ...init?.headers,
+        ...requestHeaders,
+      },
+    });
+    responseHeaders(response.headers);
+    return response;
+  };
+};
+
+describe.each(headers)(
```

Comment:
Jest expects an array of arrays here where the nth element of each element array corresponds to the nth argument of the callback function

```suggestion
describe.each(headers.map(header => [header]))(
```

Reply:
The mapping is implicit for 1-D arrays according to the docs: https://jestjs.io/docs/api#1-describeeachtablename-fn-timeout